### ISSUE TITLE
allow additional arguments to be passed to fsspec via FilesystemConfiguration

### ISSUE THREAD

**Issue by rudolfix** [dlt-hub/dlt#814]
**Background**
Our users should be able to pass additional arguments to fsspec filesystems when they are created. A good use case is to enable custom SSL checks on s3:
https://github.com/fsspec/s3fs/issues/342
requires additional argument `use_ssl` and additional element in `client_kwargs` (`verify')

**Tasks**
1. * [ ] Extent `FilesystemConfiguration` so it accepts additional arguments and client arguments ie. via config fields accepting dictionaries
2. * [ ] apply arguments in `fsspec_from_config` properly
3. * [ ] Improve the destination factory for `filesystem` to accept those additional arguments nicely, write docstrings etc.
4. * [ ] write tests that make sure that arguments are passed. ie. by providing arguments that break things in predictable way. ie. providing wrong certificate in example above
5. * [ ] PR should go to `devel` branch

**Comment by Pipboyguy** [comment]
@rudolfix Do we have a specific cert we want to use in our test suite? Does CI have some certificate installed I can use?

**Comment by rudolfix** [comment]
@Pipboyguy yeah I can create another account on AWS that requires certificate to connect. but maybe we could try this: generate a mock certificate, use it to connect and make sure that we get relevant SSL error which means that certificate is present

WDYT?

**Comment by Pipboyguy** [comment]
Sounds like the simplest solution for a negative test. However for a positive test, perhaps we can use a self signed certificate like the MinIO usecase, in which case if we use S3 we'd need a CloudFront to hand the TLS connection. Please advise

> @Pipboyguy yeah I can create another account on AWS that requires certificate to connect. but maybe we could try this: generate a mock certificate, use it to connect and make sure that we get relevant SSL error which means that certificate is present
> 
> WDYT?

**Comment by Pipboyguy** [comment]
I'll start with a local docker MinIO until further guidance

**Comment by deanja** [comment]
Following this issue as I think it will help with implementing git/github fsspec  https://github.com/dlt-hub/verified-sources/issues/301  .  

The initial case there is to pass `org` and `repo` as keyword arguments to filesystem() .  Because:
- it keeps the `bucket_url` and globbing very clean, ie `github://` and `path/to/**.md`. 
- `org` and `repo` can be included in the fsspec url as `github://org:repo@/path/to/file` but currently dlt doesn't parse those correctly.

**Comment by Pipboyguy** [comment]
This default was passed to the filesystem factory each time:

https://github.com/dlt-hub/dlt/blob/f648e9dec4f80f40c07d71ed35360e312c795540/dlt/common/storages/fsspec_filesystem.py#L95

I kept it for now, but I wonder whether this shouldn't be an additional argument passed by the user? Should it remain a default configuration?

**Comment by deanja** [comment]
Looked at the repos at https://github.com/fsspec .   It looks like many fsspec implementations - including s3fs, gdrivefs, ftp -  do cache the listings (in a DirCache) and they accept fsspec's default of `use_listings_cache = True`.   

I'm pretty new to dlt so I don't know why it is hard-coded to False.

My vote would be to leave the default as False, but give user option to set to True,

> This default was passed to the filesystem factory each time:
> 
> https://github.com/dlt-hub/dlt/blob/f648e9dec4f80f40c07d71ed35360e312c795540/dlt/common/storages/fsspec_filesystem.py#L95
> 
> I kept it for now, but I wonder whether this shouldn't be an additional argument passed by the user? Should it remain a default configuration?

**Comment by deanja** [comment]
@Pipboyguy Is it in your scope to support the additional arguments in the fsspec_filesystem() function too?  (in `dlt/common/storages/fsspec_filesystem.py`) .  That is the main way the factory gets called from the [dlt filesystem verified source](https://github.com/dlt-hub/verified-sources/blob/master/sources/filesystem/__init__.py)

**Comment by rudolfix** [comment]
@deanja the reason to get rid of the cache is that most of `dlt` operations are multiprocess or multithread so even with proper synchronization we were getting "wrong" listings due to cache. so now it is disabled by default. 

I also say explicitly that I do not want the cache when doing `ls` but I'm not sure if just `ls` is impacted or it uses cache for `exists` or `isfile` etc. `fsspec` is amazing library but the docs are lacking and code is a little bit messy so I prefer to be super strict.

**Comment by deanja** [comment]
I agree. Caches are mostly frightening!   `fsspec` cache is not accessed in most ETL use cases anyway as you:
- `ls'
- iterate once to get file details, and maybe extract content ( `fsspec` does not use cache for `open`)
- done

**Comment by Pipboyguy** [comment]
> @Pipboyguy Is it in your scope to support the additional arguments in the `fsspec_filesystem()` function too? (in `dlt/common/storages/fsspec_filesystem.py`) . That function is how the fsspec factory gets called from the [dlt filesystem verified source](https://github.com/dlt-hub/verified-sources/blob/master/sources/filesystem/__init__.py?rgh-link-date=2023-12-20T06%3A48%3A45Z)

@deanja One could argue its in this scope. This would be trivial:

```python
def fsspec_filesystem(
    protocol: str,
    credentials: FileSystemCredentials = None,
    client_kwargs: Optional[DictStrAny] = None,
    kwargs: Optional[DictStrAny] = None,
) -> Tuple[AbstractFileSystem, str]:
    return fsspec_from_config(
        FilesystemConfiguration(
            protocol, credentials, kwargs=kwargs, client_kwargs=client_kwargs
        )
    )
```
